File: src/chipschallenge/blockreactions/GreenKeyTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.Inventory.Key;

public class GreenKeyTo extends NoSlipReaction {

    private GreenKeyTo() {
    }
    private static GreenKeyTo mInstance = null;

    public static synchronized GreenKeyTo getInstance() {
        if (mInstance == null) {
            mInstance = new GreenKeyTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            takeKey(Key.GREEN);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/GreenKeyTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;
    using Type = Block.Type;

    public class GreenKeyTo : NoSlipReaction
    {
        private static GreenKeyTo instance;

        private GreenKeyTo()
        {
        }

        public static GreenKeyTo Instance
        {
            get
            {
                lock (typeof(GreenKeyTo))
                {
                    if (instance == null)
                    {
                        instance = new GreenKeyTo();
                    }
                    return instance;
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeKey(Key.GREEN);
                standing.Replace(CreateBlock(Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}